File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int sum;
    
    public boolean hasPathSum(TreeNode root, int targetSum) {
        
        if(root == null)
            return false;
        
        if(root.left == null && 
           root.right == null && 
           root.val == targetSum)
            return true;
        
        boolean hasPathSum1 = hasPathSum(root.left, targetSum - root.val);
        boolean hasPathSum2 = hasPathSum(root.right, targetSum - root.val);
        
        if(hasPathSum1 || hasPathSum2){
            return true;
        } else {
            return false;
        }
    }
}

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:



        def dfs(root, currSum):

            if not root:
                return False

            if not root.left and not root.right and currSum + root.val == targetSum:
                return True

            return dfs(root.left, currSum + root.val) or dfs(root.right, currSum + root.val) 

        if not root:
            return False

        return dfs(root, 0)                